core/templates/pages/exploration-player-page/services/exploration-engine.service.spec.ts

```diff
@@ -0,0 +1,910 @@
+      .and.returnValue(answerClassificationResult);
+
+    expect(explorationEngineService.currentStateName).toBe(undefined);
+
+    explorationEngineService.init(
+      explorationDict, 1, null, true, ['en'], initSuccessCb);
+
+    explorationEngineService.submitAnswer(
+      answer, textInputService, submitAnswerSuccessCb);
+    expect(explorationEngineService.currentStateName).toBe('Start');
+    explorationEngineService.recordNewCardAdded();
+    expect(explorationEngineService.currentStateName).toBe('Mid');
+  });
+
+  it('should load initial state when moved to new exploration', () => {
+    let moveToExplorationCb = jasmine.createSpy('success');
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+
+    explorationEngineService.exploration = explorationObjectFactory
+      .createFromBackendDict(explorationDict);
+
+    let currentStateName = explorationEngineService.currentStateName;
+    expect(currentStateName).toBe(undefined);
+
+    // Please note that we are not calling init funtion here.
+    explorationEngineService.moveToExploration(moveToExplorationCb);
+
+    currentStateName = explorationEngineService.currentStateName;
+    let initalState = explorationEngineService.exploration.initStateName;
+    expect(currentStateName).toBe(initalState);
+  });
+
+  it('should return true if current state is initial state', () => {
+    let initSuccessCb = jasmine.createSpy('success');
+
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+
+    expect(() => {
+      explorationEngineService.isCurrentStateInitial();
+    }).toThrowError('Cannot read property \'initStateName\' of undefined');
+
+    explorationEngineService.init(
+      explorationDict, 1, null, true, ['en'], initSuccessCb);
+
+    expect(explorationEngineService.isCurrentStateInitial()).toBe(true);
+  });
+
+  it('should return current state when calling \'getState\'', () => {
+    let initSuccessCb = jasmine.createSpy('success');
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+    spyOn(playerTranscriptService, 'getLastStateName').and.returnValue('Start');
+
+    expect(() => {
+      explorationEngineService.getState();
+    }).toThrowError('Cannot read property \'getState\' of undefined');
+
+    explorationEngineService.init(
+      explorationDict, 1, null, true, ['en'], initSuccessCb);
+
+    let currentState = explorationEngineService.getState();
+    expect(currentState.name).toBe('Start');
+  });
+
+  it('should return language code when calling \'getLanguageCode\'', () => {
+    let initSuccessCb = jasmine.createSpy('success');
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+
+    expect(() => {
+      explorationEngineService.getLanguageCode();
+    }).toThrowError('Cannot read property \'getLanguageCode\' of undefined');
+
+    explorationEngineService.init(
+      explorationDict, 1, null, true, ['en'], initSuccessCb);
```

Comment:
Add more input cases... This just tests with 1 possible scenario

Reply:
done.